akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala

```diff
@@ -68,10 +69,28 @@ object ServiceKey {
  * protocol spoken by that service (think of it as the set of first messages
  * that a client could send).
  */
-abstract class ServiceKey[T] extends Receptionist.AbstractServiceKey {
-  final type Protocol = T
+abstract class ServiceKey[T] extends Receptionist.AbstractServiceKey { key ⇒
+  type Protocol = T
   def id: String
   def asServiceKey: ServiceKey[T] = this
+
+  /**
+   * Scala API: Provides a type safe pattern match for listings
+   */
+  object Listing {
+    def unapply(l: Receptionist.Listing): Option[Set[ActorRef[T]]] =
+      if (l.isForKey(key)) Some(l.serviceInstances(key))
+      else None
+  }
```

Comment:
we should try this in ddata also, at least the typed ddata

Reply:
The not so nice aspect of it is ofc that the key-type needs to know about all the message types that it can be used for (making for some spaghetti dependencies between the types). But yes, very nice from a use-point of view.

I'll create a separate ticket for doing it for ddata: #24596

Reply:
> The not so nice aspect of it is ofc that the key-type needs to know about all the message types that it can be used for (making for some spaghetti dependencies between the types).

Yes, that's right. I agree that the trade-off might still be a good one.